File: clojure/bob/bob.clj
(ns bob (:require [clojure.string :as str]))

(defn silence? [s] (= (str/trim s) ""))

(defn question? [s] (= (last s) \?))

(defn yelling? [s]
  (letfn [(at-least-one-uppercase? [s]
            (->> s
                 (filter #(Character/isUpperCase %))
                 count
                 pos?))]
    (and (at-least-one-uppercase? s)
         (= (str/upper-case s) s))))

(defn response-for [s]
  (cond
    (silence? s) "Fine. Be that way!"
    (yelling? s) "Whoa, chill out!"
    (question? s) "Sure."
    :else "Whatever."))

File: elixir/bob/bob.exs
defmodule Teenager do
  def hey(input) do
    cond do
      question?(input) -> "Sure."
      shouting?(input) -> "Whoa, chill out!"
      silence?(input) -> "Fine. Be that way!"
      :else -> "Whatever."
    end
  end

  defp question?(input), do: String.ends_with?(input, "?")

  defp silence?(input), do: String.strip(input) == ""

  defp shouting?(input) do
    String.upcase(input) == input and String.downcase(input) != input
  end
end
